consensus/enclave/api/src/fee_map.rs

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! A helper object for maintaining a map of token id -> minimum fee.
+
+use alloc::{collections::BTreeMap, format, string::String};
+use core::convert::TryFrom;
+use displaydoc::Display;
+use mc_common::ResponderId;
+use mc_crypto_digestible::{DigestTranscript, Digestible, MerlinTranscript};
+use mc_sgx_compat::sync::Mutex;
+use mc_transaction_core::{constants::MINIMUM_FEE, tx::TokenId};
+use serde::{Deserialize, Serialize};
+
+/// State managed by `FeeMap`.
+struct FeeMapInner {
+    /// The actual map of token_id to fee.
+    /// Since we hash this map, it is important to use a BTreeMap as it
+    /// guarantees iterating over the map is in sorted and predictable
+    /// order.
+    pub map: BTreeMap<TokenId, u64>,
+
+    /// Cached digest value, formatted as a string.
+    /// (Suitable for appending to responder id)
+    pub cached_digest: String,
+}
+
+impl Default for FeeMapInner {
+    fn default() -> Self {
+        let mut map = BTreeMap::new();
+        map.insert(TokenId::MOB, MINIMUM_FEE);
+
+        let cached_digest = calc_digest_for_map(&map);
+
+        Self { map, cached_digest }
+    }
+}
+
+/// A thread-safe object that contains a map of fee value by token id.
+pub struct FeeMap {
+    inner: Mutex<FeeMapInner>,
+}
+
+impl Default for FeeMap {
+    fn default() -> Self {
+        Self {
+            inner: Mutex::new(FeeMapInner::default()),
+        }
+    }
+}
+
+impl TryFrom<BTreeMap<TokenId, u64>> for FeeMap {
+    type Error = Error;
+
+    fn try_from(map: BTreeMap<TokenId, u64>) -> Result<Self, Self::Error> {
+        Self::is_valid_map(&map)?;
+
+        let cached_digest = calc_digest_for_map(&map);
+
+        Ok(Self {
+            inner: Mutex::new(FeeMapInner { map, cached_digest }),
+        })
+    }
+}
+
+impl FeeMap {
+    /// Get the digest of the fee map as a hex-encoded string
+    pub fn get_digest_str(&self) -> String {
+        let inner = self.inner.lock().unwrap();
+        inner.cached_digest.clone()
+    }
+
+    /// Append the fee map digest to an existing responder id, producing a
+    /// responder id that is unique to the current fee configuration.
+    pub fn responder_id(&self, responder_id: &ResponderId) -> ResponderId {
+        ResponderId(format!("{}-{}", responder_id.0, self.get_digest_str()))
```

Comment:
The main benefit to this is avoiding the clone at line 69---like, you probably don't need the `get_digest_str()` method at all in this case.
```suggestion
        ResponderId(format!("{}-{}", responder_id.0, self.inner.lock().unwrap().cached_digest))
```

Reply:
Good call, will fix